micro_http/src/server.rs

```diff
@@ -0,0 +1,683 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+extern crate epoll;
+
+use std::io::{Error, Read, Write};
+use std::os::unix::io::AsRawFd;
+use std::os::unix::io::RawFd;
+use std::os::unix::net::{UnixListener, UnixStream};
+use std::path::Path;
+
+use common::Version;
+pub use common::{ConnectionError, RequestError};
+use connection::HttpConnection;
+use request::Request;
+use response::{Response, StatusCode};
+use std::collections::HashMap;
+
+static SERVER_FULL_ERROR_MESSAGE: &'static [u8] = b"HTTP/1.1 500\r\n\
+                                            Server: Firecracker API\r\n\
+                                            Connection: close\r\n\
+                                            Content-Length: 14\r\n\r\nServer is full";
+const MAX_CONNECTIONS: usize = 10;
+
+type Result<T> = std::result::Result<T, ServerError>;
+
+/// Errors pertaining to the server.
+#[derive(Debug)]
+pub enum ServerError {
+    /// Epoll operations failed.
+    IOError(Error),
+    /// Error from one of the connections.
+    ConnectionError(ConnectionError),
+    /// Server maximum capacity has been reached.
+    ServerFull,
+}
+
+/// Wrapper over `Request` which adds an identification token.
+pub struct RequestWithID {
+    /// Inner request.
+    pub request: Request,
+    /// Identification token.
+    id: u64,
+}
+
+impl RequestWithID {
+    /// Creates a new `RequestWithID` object from an existing `Request`,
+    /// adding an identification token.
+    pub fn new(request: Request, id: u64) -> Self {
+        RequestWithID { request, id }
+    }
+
+    /// Transforms a `RequestWithID` object into a `ResponseWithID`
+    /// retaining the identification token.
+    pub fn into_response_with_id(self, response: Response) -> ResponseWithID {
+        ResponseWithID {
+            response,
+            id: self.id,
+        }
+    }
+}
+
+/// Wrapper over `Response` which adds an identification token.
+pub struct ResponseWithID {
+    /// Inner response.
+    pub response: Response,
+    /// Identification token.
+    id: u64,
+}
+
+// Describes the state of the connection as far as data exchange
+// on the stream is concerned.
+#[derive(PartialOrd, PartialEq)]
+enum ClientConnectionState {
+    AwaitingIncoming,
+    AwaitingOutgoing,
+    Closed,
+}
+
+// Wrapper over `HttpConnection` which keeps track yielded requests
+// and absorbed responses.
+struct ClientConnection<T> {
+    // The `HttpConnection` object which handles data exchange.
+    connection: HttpConnection<T>,
+    // The state of the connection in the `epoll` structure.
+    state: ClientConnectionState,
+    // Represents the difference between yielded requests and
+    // absorbed responses.
+    // This has to be `0` if we want to drop the connection.
+    in_flight_response_count: u32,
+}
+
+impl<T: Read + Write + AsRawFd> ClientConnection<T> {
```

Comment:
`AsRawFd` not required any more for this class:
```suggestion
impl<T: Read + Write> ClientConnection<T> {
```